File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionAvg.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.math.BigDecimal;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;

public class TransformerFunctionAvg extends TransformerFunction {

    public TransformerFunctionAvg() {
        super(FunctionDescription.of(
            Map.of(
            "default", ArgumentType.of(ArgType.BigDecimal).position(0).defaultBigDecimal(BigDecimal.ZERO),
            "by", ArgumentType.of(ArgType.Transformer).position(1).defaultIsNull(true))
        ));
    }

    @Override
    public Object apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null || streamer.knownAsEmpty())
            return null;
        var hasBy = context.has("by");
        var by = context.getJsonElement( "by", false);
        var _default = Objects.requireNonNullElse(context.getBigDecimal("default"), BigDecimal.ZERO);
        var size = new AtomicInteger(0);
        var identity = BigDecimal.valueOf(0, FunctionHelpers.MAX_SCALE);
        var adapter = context.getAdapter();
        var result = streamer.stream()
                .map(t -> {
                    size.getAndIncrement();
                    var res = hasBy ? context.transformItem(by, t) : t;
                    return adapter.isNull(res) ? _default : adapter.getNumberAsBigDecimal(res);
                })
                .reduce(identity, BigDecimal::add)
                .divide(BigDecimal.valueOf(size.get()), FunctionHelpers.MAX_SCALE_ROUNDING);
        return result;
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionAvg.ts
import BigNumber from "bignumber.js";
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { isNullOrUndefined } from "../JsonHelpers";
import { BigDecimal } from "./common/FunctionHelpers";

class TransformerFunctionAvg extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        default: { type: ArgType.BigDecimal, position: 0, defaultBigDecimal: 0 },
        by: { type: ArgType.Transformer, position: 1, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const value = await context.getJsonElementStreamer(null);
    if (value == null || value.knownAsEmpty()) {
      return null;
    }
    const hasBy = context.has("by");
    const by = await context.getJsonElement("by", false);
    const _default = (await context.getBigDecimal("default")) ?? BigDecimal(0);
    let size = 0;
    const sum = await value
      .stream()
      .map(async t => {
        size++;
        const res = !hasBy ? t : await context.transformItem(by, t);
        return isNullOrUndefined(res) ? _default : BigDecimal(res);
      })
      .reduce((a: BigNumber, c) => a.plus(c));
    return sum.dividedBy(size);
  }
}

export default TransformerFunctionAvg;
